dev/source/docs/copter-adding-custom-controller.rst

```diff
@@ -0,0 +1,314 @@
+
+The custom controller library uses the same target attitude as the main controller. Most of the code inside ``AC_AttitudeControl`` library is related to input shapings such as how to interpret pilot commands depending on the flight mode or high-level position controller output. For example, in STABILIZE mode pilot RC commands are scaled based on maximum lean angle and yaw rate parameters, these values are passed to ``input_euler_angle_roll_pitch_euler_rate_yaw`` function. The pilot command is fed into a first-order input shaping algorithm to smooth out any jitter due to RC link and to generate an acceleration-limited attitude target. Later, attitude controller ``attitude_controller_run_quat`` is called and attitude error is calculated based on acceleration limited ``_attitude_target`` value, and target rate is generated. Even if acceleration limiting is turned off by setting ``ATC_RATE_FF_ENAB`` to 0, the target attitude is still presented with ``_attitude_target`` variable.
+
+Every input shaping function inside ``AC_AttitudeControl`` calls ``attitude_controller_run_quat`` at the end to run the attitude controller, except when flying in ACRO mode with the acro option set to rate only. Even in this case ``_attitude_target`` is updated properly and the pilot command can be accessed via rate_bf_targets() function.
+
+By default, the input shaping algorithm is turned on. This produces kinematically consistent attitude target and rate feed-forward values. Take a look at the custom PID backend to see how rate feed-forward is added to attitude controller output. It is advised to use the rate feed-forward value if ATC_RATE_FF_ENAB is enabled, otherwise the pilot might feel significant lag between RC command and vehicle response.
+
+After running the custom controller, mixer input is sent to the motor library via set_roll, set_pitch, set_yaw functions.
+
+
+Bumpless Transfer
+---------------------------------------------------
+
+When switching from the custom controller to the main controller it is important to reset the main controller target, error, d-term filters and set each axis integrator properly. Otherwise, a sudden jump in the controller error signal or motor output will be observed, which could result in a jerky motion. To allow a smooth transition between controllers, the main controller reset function is called when switching out of the custom controller which reset all three axis of the main controller. The reset is performed inside ``reset_main_atti_controller`` function.
+
+The attitude target and rate target are also made equal to the current attitude and gyro rate to make the error signal grow from zero in ``relax_attitude_controllers`` function. In order to avoid impulse input to the controller, the target resetting is not performed when the feedforward is disabled by setting ``ATC_RATE_FF_ENAB`` parameter to 0.
```

Comment:
```suggestion
The attitude target and rate target are also made equal to the current attitude and gyro rate to make the error signal grow from zero in ``relax_attitude_controllers`` function. In order to avoid impulse input to the controller, the target resetting is not performed when the feedforward is disabled by setting :ref:`ATC_RATE_FF_ENAB<ATC_RATE_FF_ENAB>`  parameter to 0.
```

Reply:
Done.